kernel/alarm.c

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (c) 2019 Peter Bigot Consulting, LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <logging/log.h>
+
+extern struct k_alarm _k_alarm_list_start[];
+extern struct k_alarm _k_alarm_list_end[];
+
+static struct k_spinlock lock;
+static sys_dlist_t scheduled_list = SYS_DLIST_STATIC_INIT(&scheduled_list);
+static sys_dlist_t ready_list = SYS_DLIST_STATIC_INIT(&ready_list);
+
+/** Return a pointer to the k_alarm that contains the node if it has a
+ * deadline at or before the provided when.  Return a null pointer if
+ * the k_alarm is scheduled strictly after the provided when.
+ */
+static ALWAYS_INLINE struct k_alarm *alarm_due_by(sys_dnode_t *node,
+						  uint32_t when)
+{
+	struct k_alarm *cap = CONTAINER_OF(node, struct k_alarm, node);
+
+	if ((s32_t)(cap->deadline - when) > 0) {
+		return 0;
+	}
+	return cap;
+}
+
+/** Add alarm to the provided list before the first alarm that has a
+ * later deadline, or at the end if necessary.
+ *
+ * Return @true if the head of list changed as a result of this
+ * addition.
+ */
+static bool link_alarm(sys_dlist_t *list,
+		       struct k_alarm *alarm,
+		       u32_t now)
+{
+	sys_dnode_t *node = &alarm->node;
+	sys_dnode_t *cp = sys_dlist_peek_head(list);
+
+	while (cp) {
+		if (!alarm_due_by(cp, alarm->deadline)) {
+			break;
+		}
+		cp = sys_dlist_peek_next_no_check(list, cp);
+	}
+	if (cp) {
+		sys_dlist_insert(cp, node);
+	} else {
+		sys_dlist_append(list, node);
+	}
+	return sys_dlist_is_head(list, node);
+}
+
+void k_alarm_init(struct k_alarm *alarm,
+		  k_alarm_handler_t handler,
+		  k_alarm_cancel_t cancel,
+		  void *user_data)
+{
+	*alarm = Z_ALARM_INITIALIZER(handler, cancel, user_data);
+}
+
+int k_alarm_next_deadline_(u32_t *deadline)
+{
+	int rv = -1;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	if (!sys_dlist_is_empty(&ready_list)) {
+		rv = 0;
+		goto unlock;
+	}
+
+	sys_dnode_t *hp = sys_dlist_peek_head(&scheduled_list);
```

Comment:
Never declare a variable in the middle of a code block.

Reply:
Can you elaborate where this "never" comes from? This uninitialised variable safety issue was fixed in C99, twenty years ago.